pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java

```diff
@@ -90,8 +90,24 @@ public BlobInputStream(LargeObject lo, int bsize, long limit) {
     // The very first read multiplies the last buffer size by two, so we divide by two to get
     // the first read to be exactly the initial buffer size
     this.lastBufferSize = INITIAL_BUFFER_SIZE / 2;
+
+    try {
+      // initialise current position for mark/reset
+      this.absolutePosition = lo.tell64();
```

Comment:
So we have access to the baseConnection through `lo.fp.connection` we can get the server version from there and avoid this dance.

Reply:
We could avoid the try/catch, yes, but we still need to do the `tell` to fix #3149.

Reply:
agreed. As much as I dislike putting the call in the constructor I don't think it makes that much of a difference.

Reply:
After contemplating this some more, I think it makes more sense to put the tell in getLo(). We expect read/write move operations to throw exceptions, we do not expect constructors to .

Reply:
But the constructor here is public. This class has to be able to deal with a passed lo that is not at the start.

I don't see why constructors cannot throw, but (as previously discussed) the only other solution is to defer this check to the first time any other method is called, which makes for brittle and repetitive code.

Reply:
It would be very surprising to me if my java program threw a `RuntimeException` because of a connection error. If you put it in getLo() which is called by every read/write/move operation it would only be in one place, there would not be duplicate or repetitive code and we expect a SQLException here .

Reply:
Oh I see, sorry I forgot where that method was. OK, I'll look at doing that.

Reply:
Sorry, that doesn't work. `mark(int)` is not allowed to throw any checked exceptions, but needs to have had the position initialized.

Reply:
Also "we have access to the baseConnection" isn't true. Those fields are private.

Reply:
Looking at mark, we don't even implement that correctly as we ignore the readLimit. I'd be fine with ignoring the exception there. After all you can call mark on a closed input stream.

Reply:
There are no issues with the `readLimit`. The mark position is never invalidated by reading. It is there to allow a stream to allocate a buffer when its underlying source cannot seek backwards.

Yes, ignoring exceptions there should be OK, though I'll still log it so that the cause of the next exception is not completely lost. I should have that done tomorrow.

Reply:
I do not like the idea of throwing an exception on every `lo.tell64` call.

I would rather suggest implementing `int FastPath#getServerVersionNum()`, so `LargeObject#tell64` could automatically call `lo_tell` if the database does not support tell64.

Reply:
That sounds terrible. `LargeObject#tell64` should always call `tell64` and `LargeObject#tell` should always call `tell`.

Reply:
Please come up with the PR that fulfills the following then:
* it should work with the supported DB versions
* it should not degrade performance
* it should support LO exceeding MAX_INT

Reply:
That is what this PR does...

Reply:
I'm afraid the PR "degrades the performance for PG<9.3" since it would always try `tell64`, catch the exception and try `tell`. Would you please be so kind to revalidate and ensure the PR fulfills the above requirements?

Reply:
This PR massively improves the performance for all postgres versions by allowing it to actually seek instead of having to fetch all the data it's trying to skip over.

The single additional call I have added for the bug fix is unavoidable if you want the behaviour to be correct. The options are always two call (check version, then use correct tell), or only one call for the majority of users (try tell64, then tell if it fails).